--- network/jsclient/ApiClient.js	2018-12-13 12:16:14.874596171 +0100
+++ network/jsclient/ApiClient.js	2018-12-13 12:05:13.798575996 +0100
@@ -174,16 +174,6 @@
    * @returns {Boolean} <code>true</code> if <code>param</code> represents a file.
    */
   exports.prototype.isFileParam = function(param) {
-    // fs.ReadStream in Node.js and Electron (but not in runtime like browserify)
-    if (typeof require === 'function') {
-      var fs;
-      try {
-        fs = require('fs');
-      } catch (err) {}
-      if (fs && fs.ReadStream && param instanceof fs.ReadStream) {
-        return true;
-      }
-    }
     // Buffer in Node.js
     if (typeof Buffer === 'function' && param instanceof Buffer) {
       return true;
